sdk/python/kfp/dsl/pipeline_channel.py

```diff
@@ -265,6 +284,234 @@ def __init__(
+        from kfp.dsl import for_loop
+        if any(isinstance(channel, for_loop.Collected) for channel in channels):
+            raise ValueError(
+                f'dsl.{for_loop.Collected.__name__} cannot be used inside of dsl.{OneOf.__name__}.'
+            )
+
+    def _validate_no_oneof_channel(
+        self, channels: List[Union[PipelineParameterChannel,
+                                   PipelineArtifactChannel]]
+    ) -> None:
+        if any(isinstance(channel, OneOfMixin) for channel in channels):
+            raise ValueError(
+                f'dsl.{OneOf.__name__} cannot be used inside of another dsl.{OneOf.__name__}.'
+            )
+
+    def _validate_no_mix_of_parameters_and_artifacts(
+        self, channels: List[Union[PipelineParameterChannel,
+                                   PipelineArtifactChannel]]
+    ) -> None:
+        readable_name_map = {
+            PipelineParameterChannel: 'parameter',
+            PipelineArtifactChannel: 'artifact',
+            OneOfParameter: 'parameter',
+            OneOfArtifact: 'artifact',
+        }
+        # if channels[0] is any subclass of a PipelineParameterChannel
+        # check the rest of the channels against that parent type
+        # ensures check permits OneOfParameter and PipelineParameterChannel
+        # to be passed to OneOf together
+        if isinstance(channels[0], PipelineParameterChannel):
+            expected_type = PipelineParameterChannel
+        else:
+            expected_type = PipelineArtifactChannel
+
+        for i, channel in enumerate(channels[1:], start=1):
+            if not isinstance(channel, expected_type):
+                raise TypeError(
+                    f'Task outputs passed to dsl.{OneOf.__name__} must be the same type. Got two channels with different types: {readable_name_map[expected_type]} at index 0 and {readable_name_map[type(channel)]} at index {i}.'
+                )
+
+    def _validate_has_else_group(
+        self,
+        parent_group: 'tasks_group.ConditionBranches',
+    ) -> None:
+        # avoid circular imports
+        from kfp.dsl import tasks_group
+        if not isinstance(parent_group.groups[-1], tasks_group.Else):
+            raise ValueError(
+                f'dsl.{OneOf.__name__} must include an output from a task in a dsl.{tasks_group.Else.__name__} group to ensure at least one output is available at runtime.'
+            )
+
+    def __str__(self):
+        # supporting oneof in f-strings is technically feasible, but would
+        # require somehow encoding all of the oneof channels into the
+        # f-string
+        # another way to do this would be to maintain a pipeline-level
+        # map of PipelineChannels and encode a lookup key in the f-string
+        # the combination of OneOf and an f-string is not common, so prefer
+        # deferring implementation
+        raise NotImplementedError(
+            f'dsl.{OneOf.__name__} is not yet supported in f-strings.')
```

Comment:
nit: Technically `f-string` is just one special syntax for string concatenation, while this limitation would apply to all formats, for example, `"{} world".format(str1)`, `"Hello" + str1`, etc.